### PR TITLE
Speed up dec2flt fast path with additional tables.

### PR DESCRIPTION
**Pull request by hanna-kruppe**
Add tables of small powers of ten used in the fast path. The tables are redundant: We could also use the big, more accurate table and round the value to the correct type (in fact we did just that before this commit). However, the rounding is extra work and slows down the fast path.

Because only very small exponents enter the fast path, the table and thus the space overhead is negligible. Speed-wise, this is a clear win on a [benchmark](https://gist.github.com/Veedrac/dbb0c07994bc7882098e) comparing the fast path to a naive, hand-optimized, inaccurate algorithm. Specifically, this change narrows the gap from a roughly 5x difference to a roughly 3.4x difference.

### CODE REVIEW

**Comment by rkruppe** [comment]
Part of #30638

**Comment by alexcrichton** [comment]
@bors: r+ 5cd98e0913f5f6165fb55097ad2861f8cdc3b2fb

Nice!

File: src/libcore/num/dec2flt/algorithm.rs [hunk 1]
```diff
@@ -60,17 +60,13 @@ pub fn fast_path<T: RawFloat>(integral: &[u8], fractional: &[u8], e: i64) -> Opt
...
-    // The case `e == 0`, however, is unnecessary for correctness. It's just measurably faster.
-    if e == 0 {
-        Some(T::from_int(f))
-    } else if e > 0 {
-        Some(T::from_int(f) * fp_to_float(power_of_ten(e)))
+    if e > 0 {
+        Some(T::from_int(f) * T::short_fast_pow10(e as usize))
     } else {
-        Some(T::from_int(f) / fp_to_float(power_of_ten(-e)))
+        Some(T::from_int(f) / T::short_fast_pow10((-e) as usize))
```
Comment (pnkfelix) [inline src/libcore/num/dec2flt/algorithm.rs]: Wow is `ident as type` really more tightly binding than negation in our grammar?   Or did you just find this easier to read?  

(My eyes kept folding the two opening parens into just one,  leading me to think the cast was applied to the result of the call)

Comment (rkruppe) [inline src/libcore/num/dec2flt/algorithm.rs]: I simply don't know which binds more tightly. I checked now and it seems `-e as usize` would work, but ask me tomorrow and I'll be unsure again. I should just use `e.abs()`. r- please

**Comment by rkruppe** [comment]
Failure seems unrelated. I want to fix the style thing @pnkfelix pointed out so no need to retry immediately.

File: src/libcore/num/dec2flt/algorithm.rs (hunk 1 above)
Comment (rkruppe) [inline src/libcore/num/dec2flt/algorithm.rs]: Done. If the Playground did not deceive me, `e.abs()` should even generate identical code.

**Pull request merged**